### DISCUSSION TITLE
CosineTransformer does not implement inverse_transform

### DISCUSSION THREAD

**Discussion by raul-torres-diaz** [alan-turing-institute/sktime#3662] [Q&A] [answered]
Hi all,
I'm running a transformer pipeline and when I try to get the inverse, it is rising an error and checking the elements of the pipeline I dicscover that it was due to the CosineTransformer() step.

```
cos = CosineTransformer()
train = cos.fit_transform(df_train)
res = cos.inverse_transform(test)
```

Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2022.2.2\plugins\python-ce\helpers\pydev\_pydevd_bundle\pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
  File "C:\Users\ratorres\Documents\Work_projects\IA4HA\PT2\signal_event_detection\venv\lib\site-packages\sktime\transformations\base.py", line 592, in inverse_transform
    raise NotImplementedError(
NotImplementedError: <class 'sktime.transformations.series.cos.CosineTransformer'> does not implement inverse_transform

**Comment by fkiraly** [comment]
Hm, in the code it's turned off, since `cos` is not invertible outside pi/-pi.

Should we turn this on? Probably some tests will faill when data outside the pi/-pi interval is generated, but maybe we should just ensure the users don't pass data outside the range...

**Reply by fkiraly** [reply]
PR here to see what breaks:
https://github.com/sktime/sktime/pull/3671

**Reply by Douger68** [reply]
This appears to not have been turned on or fixed.  Can someone look into it?

**Reply by fkiraly** [reply]
hm, @Douger68, @raul-torres-diaz, may I ask what is the expectation that it should do?

* That the inverse produces inverses for the invertible range?
* that the inverse passes through values?

The tests break since the transformer is tested on values outside its invertible range - e.g., it does not recover original values as the contract assumes. This could be resolved with test skips or contract relaxation.

**Comment by fkiraly** [comment, marked as answer]
I upgraded to PR to deal with invertibility in a limited range. This should now allow to use the `CosineTransformer` with compositors that require an inverse: https://github.com/sktime/sktime/pull/3671

Caveat, if values passed to `transform` fall outside the range, the inverse will end up within the range.

**Reply by fkiraly** [reply]
@raul-torres-diaz, @Douger68, this is now merged, so should be in the next release.

If you want to test, you can try the current latest developmet version.
